README.md
# go-multipart

[![CI](https://github.com/030/multipart/workflows/Go/badge.svg?event=push)](https://github.com/030/multipart/actions?query=workflow%3AGo)
[![GoDoc Widget]][godoc]
![GitHub go.mod Go version](https://img.shields.io/github/go-mod/go-version/030/multipart?logo=go)
[![Go Report Card](https://goreportcard.com/badge/github.com/030/multipart)](https://goreportcard.com/report/github.com/030/multipart)
[![StackOverflow SE Questions](https://img.shields.io/stackexchange/stackoverflow/t/multipart.svg?logo=stackoverflow)](https://stackoverflow.com/tags/multipart)
[![DevOps SE Questions](https://img.shields.io/stackexchange/devops/t/multipart.svg?logo=stackexchange)](https://devops.stackexchange.com/tags/multipart)
[![ServerFault SE Questions](https://img.shields.io/stackexchange/serverfault/t/multipart.svg?logo=serverfault)](https://serverfault.com/tags/multipart)
[![Docker Pulls](https://img.shields.io/docker/pulls/utrecht/multipart?logo=docker&logoColor=white)](https://hub.docker.com/r/utrecht/multipart)
![Docker Image Size (latest semver)](https://img.shields.io/docker/image-size/utrecht/multipart?logo=docker&logoColor=white&sort=semver)
![Issues](https://img.shields.io/github/issues-raw/030/multipart.svg)
![Pull requests](https://img.shields.io/github/issues-pr-raw/030/multipart.svg)
![Total downloads](https://img.shields.io/github/downloads/030/multipart/total.svg)
![GitHub forks](https://img.shields.io/github/forks/030/multipart?label=fork&style=plastic)
![GitHub watchers](https://img.shields.io/github/watchers/030/multipart?style=plastic)
![GitHub stars](https://img.shields.io/github/stars/030/multipart?style=plastic)
![License](https://img.shields.io/github/license/030/multipart.svg)
![Repository Size](https://img.shields.io/github/repo-size/030/multipart.svg)
![Contributors](https://img.shields.io/github/contributors/030/multipart.svg)
![Commit activity](https://img.shields.io/github/commit-activity/m/030/multipart.svg)
![Last commit](https://img.shields.io/github/last-commit/030/multipart.svg)
![Release date](https://img.shields.io/github/release-date/030/multipart.svg)
[![GitHub release (latest SemVer)](https://img.shields.io/github/v/release/030/multipart?logo=github&sort=semver)](https://github.com/030/multipart/releases/latest)
[![Bugs](https://sonarcloud.io/api/project_badges/measure?project=030_multipart&metric=bugs)](https://sonarcloud.io/dashboard?id=030_multipart)
[![Code Smells](https://sonarcloud.io/api/project_badges/measure?project=030_multipart&metric=code_smells)](https://sonarcloud.io/dashboard?id=030_multipart)
[![Coverage](https://sonarcloud.io/api/project_badges/measure?project=030_multipart&metric=coverage)](https://sonarcloud.io/dashboard?id=030_multipart)
[![Duplicated Lines (%)](https://sonarcloud.io/api/project_badges/measure?project=030_multipart&metric=duplicated_lines_density)](https://sonarcloud.io/dashboard?id=030_multipart)
[![Lines of Code](https://sonarcloud.io/api/project_badges/measure?project=030_multipart&metric=ncloc)](https://sonarcloud.io/dashboard?id=030_multipart)
[![Maintainability Rating](https://sonarcloud.io/api/project_badges/measure?project=030_multipart&metric=sqale_rating)](https://sonarcloud.io/dashboard?id=030_multipart)
[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=030_multipart&metric=alert_status)](https://sonarcloud.io/dashboard?id=030_multipart)
[![Reliability Rating](https://sonarcloud.io/api/project_badges/measure?project=030_multipart&metric=reliability_rating)](https://sonarcloud.io/dashboard?id=030_multipart)
[![Security Rating](https://sonarcloud.io/api/project_badges/measure?project=030_multipart&metric=security_rating)](https://sonarcloud.io/dashboard?id=030_multipart)
[![Technical Debt](https://sonarcloud.io/api/project_badges/measure?project=030_multipart&metric=sqale_index)](https://sonarcloud.io/dashboard?id=030_multipart)
[![Vulnerabilities](https://sonarcloud.io/api/project_badges/measure?project=030_multipart&metric=vulnerabilities)](https://sonarcloud.io/dashboard?id=030_multipart)
[![codecov](https://codecov.io/gh/030/multipart/branch/main/graph/badge.svg)](https://codecov.io/gh/030/multipart)
[![GolangCI](https://golangci.com/badges/github.com/golangci/golangci-web.svg)](https://golangci.com/r/github.com/030/multipart)
[![Chocolatey](https://img.shields.io/chocolatey/dt/multipart)](https://chocolatey.org/packages/multipart)
[![multipart](https://snapcraft.io//multipart/badge.svg)](https://snapcraft.io/multipart)
[![codebeat badge](https://codebeat.co/badges/f4aa5086-a4d5-41cd-893a-5da816ee9107)](https://codebeat.co/projects/github-com-030-multipart-main)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-%23FE5196?logo=conventionalcommits&logoColor=white)](https://conventionalcommits.org)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

[godoc]: https://godoc.org/github.com/030/multipart
[godoc widget]: https://godoc.org/github.com/030/multipart?status.svg

## Quickstart

Start a Nexus3 server:

```bash
docker run --rm -d -p 9000:8081 sonatype/nexus3:3.16.0
```

Upload artifacts once it has been started:

```bash
./multipart -url \
    http://localhost:9000/service/rest/v1/components?repository=maven-releases \
    -user admin -pass admin123 -F \
    "maven2.asset1=@test/testdata/file1.pom,\
    maven2.asset1.extension=pom,\
    maven2.asset2=@test/testdata/file1.jar,\
    maven2.asset2.extension=jar,\
    maven2.asset3=@test/testdata/file1.war,\
    maven2.asset3.extension=war,\
    maven2.asset4=@test/testdata/file1.ear,\
    maven2.asset4.extension=ear,\
    maven2.asset5=@test/testdata/file1-sources.jar,\
    maven2.asset5.classifier=sources,\
    maven2.asset5.extension=jar"
```

[godoc]: https://godoc.org/github.com/030/go-multipart
[godoc widget]: https://godoc.org/github.com/030/go-multipart?status.svg


docs/CHANGELOG.md
<a name="unreleased"></a>
## [Unreleased]


<a name="v1.0.7"></a>
## [v1.0.7] - 2023-01-15
### Build
- **deps:** Update versions.


<a name="v1.0.6"></a>
## [v1.0.6] - 2022-12-16
### Build
- **auto-updater:** Update schedule.


<a name="v1.0.5"></a>
## [v1.0.5] - 2022-11-13

<a name="v1.0.4"></a>
## [v1.0.4] - 2022-11-09
### Fix
- resolve issues in structure.
- Add autoupdater.
- Use codecov v3 action to resolve current issues in uploading.

### BREAKING CHANGE

This project should be imported as multipart.


<a name="v1.0.3"></a>
## [v1.0.3] - 2022-11-08
### Fix
- Solve issues in structure.


<a name="v1.0.2"></a>
## [v1.0.2] - 2020-03-24

<a name="v1.0.1"></a>
## [v1.0.1] - 2019-10-20
### Reverts
- [[GH-16](https://github.com/030/multipart/issues/16)] Try to solve 'diff: /dev/fd/63: No such file or directory' on windows travis


<a name="v1.0.0"></a>
## v1.0.0 - 2019-05-26

[Unreleased]: https://github.com/030/multipart/compare/v1.0.7...HEAD
[v1.0.7]: https://github.com/030/multipart/compare/v1.0.6...v1.0.7
[v1.0.6]: https://github.com/030/multipart/compare/v1.0.5...v1.0.6
[v1.0.5]: https://github.com/030/multipart/compare/v1.0.4...v1.0.5
[v1.0.4]: https://github.com/030/multipart/compare/v1.0.3...v1.0.4
[v1.0.3]: https://github.com/030/multipart/compare/v1.0.2...v1.0.3
[v1.0.2]: https://github.com/030/multipart/compare/v1.0.1...v1.0.2
[v1.0.1]: https://github.com/030/multipart/compare/v1.0.0...v1.0.1


integration-tests.sh
#!/bin/bash -e

TOOL="${1:-go run main.go}"

validate(){
    if [ -z "$TOOL" ]; then
        echo "No deliverable defined. Assuming that 'go run main.go' 
should be run."
        TOOL="./go-multipart"
    fi
}

nexus(){
    docker run --rm -d -p 9999:8081 --name nexus sonatype/nexus3:3.16.1
}

readiness(){
    until docker logs nexus | grep 'Started Sonatype Nexus OSS'
    do
        echo "Nexus unavailable"
        sleep 10
    done
}

unit(){
    f=cover.out; if [ -f $f ]; then rm $f; fi; go test ./... -coverprofile $f && go tool cover -html $f
}

artifacts(){
    $TOOL -url \
    http://localhost:9999/service/rest/v1/components?repository=maven-releases \
    -user admin -pass admin123 -F \
    "maven2.asset1=@utils/test-files-multipart/file1.pom,\
    maven2.asset1.extension=pom,\
    maven2.asset2=@utils/test-files-multipart/file1.jar,\
    maven2.asset2.extension=jar,\
    maven2.asset3=@utils/test-files-multipart/file1-sources.jar,\
    maven2.asset3.extension=sources.jar"
}

cleanup(){
    docker stop nexus
}

main(){
    validate
    nexus
    readiness
    artifacts
    unit
}

trap cleanup EXIT
main

utils/multipart_test.go
package utils

import (
	"reflect"
	"testing"
)

const expPostErr string = "Post \"\": unsupported protocol scheme \"\""

func TestMultipartUpload(t *testing.T) {
	u := Upload{URL: "http://localhost:9999/service/rest/v1/components?repository=maven-releases", Username: "admin", Password: "admin123"}

	// happy
	err := u.MultipartUpload("maven2.asset1=@test-files-multipart/file2.pom,maven2.asset1.extension=pom")
	if err != nil {
		t.Errorf("Error should be 'nil', but was '%v'", err)
	}

	// unhappy
	err = u.MultipartUpload("hello")
	want := "The string should at least contain a '=', but was: 'hello'"
	if err != nil {
		if err.Error() != want {
			t.Errorf("An error was expected. Got: '%v', want: '%s'", err, want)
		}
	}

	// unhappy - upload should fail if endpoint is incorrect
	u = Upload{URL: "", Username: "", Password: ""}
	err = u.MultipartUpload("maven2.asset1.extension=pom")
	want = expPostErr
	if err != nil {
		if err.Error() != want {
			t.Errorf("An error was expected. Got: '%v', want: '%s'", err, want)
		}
	}
}

func TestUpload(t *testing.T) {
	u := Upload{URL: "", Username: "", Password: ""}
	err := u.upload()
	want := expPostErr
	if err.Error() != want {
		t.Errorf("An error was expected. Received '%v'; want '%s'", err, want)
	}
}

func TestStringToSlice(t *testing.T) {
	testMap := map[string][]string{
		"":                []string{""},
		"hello,world":     []string{"hello", "world"},
		"hello,world,ola": []string{"hello", "world", "ola"},
	}

	for k, v := range testMap {
		got := stringToSlice(k)
		want := v
		if !reflect.DeepEqual(want, got) {
			t.Errorf("Slice not identical. Expected %s, but was %s.", want, got)
		}
	}
}

func TestReadFile(t *testing.T) {
	_, err := readFile("abc")
	want := "open abc: no such file or directory"
	if err.Error() != want {
		t.Errorf("An error was expected. Got '%v', want '%s'", err, want)
	}
}

func TestMultipartBody(t *testing.T) {
	in := []string{"this is invalid input"}

	err := multipartBody(in)
	want := "The string should at least contain a '=', but was: 'this is invalid input'"

	if err.Error() != want {
		t.Errorf("An error was expected. Received '%v'; want '%s'", err, want)
	}
}


pkg/multipart/multipart_test.go
package multipart

import (
	"os"
	"testing"

	"github.com/030/mij"
	"github.com/030/multipart/internal/app/multipart/multiparttest"
)

const expPostErr string = "Post \"\": unsupported protocol scheme \"\""

func TestMain(m *testing.M) {
	containers := []mij.DockerImage{multiparttest.Image(10000)}
	if err := multiparttest.Setup(containers); err != nil {
		panic(err)
	}

	code := m.Run()
	if err := multiparttest.Shutdown(containers); err != nil {
		panic(err)
	}

	os.Exit(code)
}

var u = Upload{URL: "http://localhost:10000/service/rest/v1/components?repository=maven-releases", Username: "admin", Password: "admin123"}

func TestUpload(t *testing.T) {
	err := u.Upload("maven2.asset1=@../../test/testdata/file2.pom,maven2.asset1.extension=pom")
	if err != nil {
		t.Errorf("Error should be 'nil', but was '%v'", err)
	}
}

func TestUploadFail(t *testing.T) {
	err := u.Upload("hello")
	want := "the string should at least contain a '=', but was: 'hello'"
	if err != nil {
		if err.Error() != want {
			t.Errorf("An error was expected. Got: '%v', want: '%s'", err, want)
		}
	}

	u = Upload{URL: "", Username: "", Password: ""}
	err = u.Upload("maven2.asset1.extension=pom")
	want = expPostErr
	if err != nil {
		if err.Error() != want {
			t.Errorf("An error was expected. Got: '%v', want: '%s'", err, want)
		}
	}
}


pkg/multipart/multipart.go
package multipart

import (
	"bytes"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"

	"github.com/030/multipart/internal/app/multipart/utils"
)

// Upload struct.
type Upload struct {
	URL      string
	Username string
	Password string
}

var (
	body   = new(bytes.Buffer)
	writer = multipart.NewWriter(body)
)

func (u Upload) request() (*http.Request, error) {
	req, err := http.NewRequest("POST", u.URL, body)
	if err != nil {
		return nil, err
	}

	req.Header.Add("Content-Type", writer.FormDataContentType())
	req.SetBasicAuth(u.Username, u.Password)

	return req, nil
}

func response(req *http.Request) (*http.Response, error) {
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}

	return resp, nil
}

// MultipartUpload splits the string into a slice, creates a multipart
// and that is posted to an URL.
func (u Upload) Upload(s string) error {
	a := utils.StringToSlice(s)
	if err := utils.Body(writer, a); err != nil {
		return fmt.Errorf("%w", err)
	}

	req, err := u.request()
	if err != nil {
		return err
	}

	resp, err := response(req)
	if err != nil {
		return err
	}
	defer func() {
		if err := resp.Body.Close(); err != nil {
			panic(err)
		}
	}()

	b, err := io.ReadAll(resp.Body)
	if !(resp.StatusCode == http.StatusOK || resp.StatusCode == http.StatusNoContent) || (err != nil) {
		return fmt.Errorf("HTTPStatusCode: '%d'; ResponseMessage: '%s'; ErrorMessage: '%w'", resp.StatusCode, string(b), err)
	}

	return nil
}


internal/app/multipart/multiparttest/docker.go
package multiparttest

import (
	"strconv"
	"sync"

	"github.com/030/mij"
	log "github.com/sirupsen/logrus"
)

func Setup(containers []mij.DockerImage) error {
	log.SetLevel(log.DebugLevel)
	var wg sync.WaitGroup
	for _, container := range containers {
		wg.Add(1)
		go func(m mij.DockerImage) {
			defer wg.Done()
			if err := m.Run(); err != nil {
				panic(err)
			}
		}(container)
	}
	wg.Wait()

	return nil
}

func Shutdown(containers []mij.DockerImage) error {
	var wg sync.WaitGroup
	for _, container := range containers {
		wg.Add(1)
		go func(m mij.DockerImage) {
			defer wg.Done()
			if err := m.Stop(); err != nil {
				panic(err)
			}
		}(container)
	}
	wg.Wait()

	return nil
}

func Image(port int) mij.DockerImage {
	return mij.DockerImage{
		Name:                     "sonatype/nexus3",
		PortExternal:             port,
		PortInternal:             8081,
		Version:                  "3.16.0",
		ContainerName:            "nexus" + strconv.Itoa(port),
		LogFile:                  "/nexus-data/log/nexus.log",
		LogFileStringHealthCheck: "Started Sonatype Nexus OSS",
	}
}
